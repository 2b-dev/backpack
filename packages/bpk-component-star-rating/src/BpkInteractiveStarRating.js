import React from 'react';
import PropTypes from 'prop-types';

import { STAR_TYPES } from './BpkStar';
import BpkInteractiveStar from './BpkInteractiveStar';
import './bpk-star-rating.scss';

const getTypeByRating = (starNumber, rating) => {
  if (starNumber > rating) {
    return STAR_TYPES.EMPTY;
  }

  return STAR_TYPES.FULL;
};

const BpkStarRating = (props) => {
  const {
    className,
    getStarLabel,
    hoverRating,
    id,
    large,
    maxRating,
    onMouseLeave,
    onRatingHover,
    onRatingSelect,
    rating,
    ...rest
  } = props;

  const stars = [];
  const classNames = ['bpk-star-rating'];
  const displayRating = hoverRating || rating;

  const currentRating = displayRating > maxRating ? maxRating : displayRating;

  if (className) { classNames.push(className); }

  for (let starNumber = 1; starNumber <= maxRating; starNumber += 1) {
    const type = getTypeByRating(starNumber, currentRating);

    stars.push(
      <BpkInteractiveStar
        key={`star${starNumber}`}
        onChange={event => onRatingSelect(starNumber, event)}
        type={type}
        large={large}
        onMouseEnter={() => onRatingHover(starNumber)}
        selected={rating === starNumber}
        label={getStarLabel(starNumber, maxRating)}
        name={`${id}_rating`}
        value={starNumber}
      />,
    );
  }

  return (
    <div
      {...rest}
      className={classNames.join(' ')}
      onMouseLeave={onMouseLeave}
    >
      {stars}
    </div>
  );
};

BpkStarRating.propTypes = {
  id: PropTypes.string.isRequired,
  getStarLabel: PropTypes.func.isRequired,
  rating: PropTypes.number,
  hoverRating: PropTypes.number,
  maxRating: PropTypes.number,
  large: PropTypes.bool,
  onRatingHover: PropTypes.func,
  onMouseLeave: PropTypes.func,
  onRatingSelect: PropTypes.func,
  className: PropTypes.string,
};

BpkStarRating.defaultProps = {
  rating: 0,
  hoverRating: 0,
  maxRating: 5,
  large: false,
  onRatingHover: () => null,
  onMouseLeave: () => null,
  onRatingSelect: () => null,
  className: null,
};

export default BpkStarRating;
