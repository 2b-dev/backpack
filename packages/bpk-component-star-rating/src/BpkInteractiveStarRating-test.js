import React from 'react';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import BpkInteractiveStarRating from './BpkInteractiveStarRating';

describe('BpkInteractiveStarRating', () => {
  it('should render correctly', () => {
    const tree = shallow(<BpkInteractiveStarRating />);
    expect(toJson(tree)).toMatchSnapshot();
  });

  it('should render correctly with "large" attribute', () => {
    const tree = shallow(<BpkInteractiveStarRating large />);
    expect(toJson(tree)).toMatchSnapshot();
  });

  it('should have a rating of 3 when onRatingSelect is called with 3', () => {
    const interactiveItem = shallow(<BpkInteractiveStarRating />);
    interactiveItem.instance().onRatingSelect(3);
    expect(interactiveItem.state('rating')).toBe(3);
  });

  it('should have a hover rating of 4 when onRatingHover is called with 4', () => {
    const interactiveItem = shallow(<BpkInteractiveStarRating />);

    interactiveItem.instance().onRatingHover(4);
    expect(interactiveItem.state('hoverRating')).toBe(4);

    interactiveItem.instance().onMouseLeave();
    expect(interactiveItem.state('hoverRating')).toBe(0);
  });
});
