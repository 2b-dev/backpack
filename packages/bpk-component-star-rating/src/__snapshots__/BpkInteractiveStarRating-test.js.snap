exports[`BpkInteractiveStarRating should render correctly 1`] = `
<BpkStarRating
  className={null}
  hoverRating={0}
  interactive={true}
  large={false}
  maxRating={5}
  onMouseLeave={[Function]}
  onRatingHover={[Function]}
  onRatingSelect={[Function]}
  rating={0} />
`;

exports[`BpkInteractiveStarRating should render correctly with "large" attribute 1`] = `
<BpkStarRating
  className={null}
  hoverRating={0}
  interactive={true}
  large={true}
  maxRating={5}
  onMouseLeave={[Function]}
  onRatingHover={[Function]}
  onRatingSelect={[Function]}
  rating={0} />
`;
